game/view/actor/attr.lua

```diff
@@ -0,0 +1,148 @@
+  particles:setEmissionArea('ellipse', 0, 0, 0, false)
+  particles:setTangentialAcceleration(-512)
+  return particles
+end
+
+local function _renderAttribute(data)
+  local g = love.graphics
+  local actor, attrname, particles = unpack(data)
+  local progress = 0
+  local lvl = actor:getAttrLevel(attrname)
+  while true do
+    -- update bar progress
+    local upgrade = actor:getAttrUpgrade(attrname)
+    local aptitude = actor:getAptitude(attrname)
+    local total_prev = APT.CUMULATIVE_REQUIRED_ATTR_UPGRADE(aptitude, lvl)
+    local total_next = APT.CUMULATIVE_REQUIRED_ATTR_UPGRADE(aptitude, lvl+1)
+    local target = min(1 ,
+      max(0, (upgrade - total_prev) / (total_next - total_prev))
+    )
+    progress = min(
+      1, max(0, progress + 8*(target-progress)*_dt)
+    )
+    if abs(target - progress) < 0.01 then
+      progress = target
+    end
+
+    -- check if level up
+    local newlvl = actor:getAttrLevel(attrname)
+    local offset = newlvl - lvl
+    local step = offset / abs(offset)
+    if newlvl ~= lvl and progress == 1 then
+      lvl = lvl + step
+      offset = offset - step
+      progress = 0
+      particles:emit(48)
+      PLAYSFX('get-item')
+      local start = 0
+      local rand = RANDOM.safeGenerate
+      while start <= 0.5 do
+        g.translate(2*(rand()*2-1), 2*(rand()*2-1))
+        start = start + _dt
+        coyield(lvl, offset, progress, true)
+      end
+    end
+
+    -- yield
+    coyield(lvl, offset, progress)
+  end
+end
+
+local ATTR = {}
+
+function ATTR.init(width)
+  _barwidth = width/4
+  _font = FONT.get("Text", 20)
+end
+
+function ATTR.draw(g, actor, attrname)
+  g.push()
+  _dt = delta()
+  local actorstate = _states[actor] or {}
+  local attrstate = actorstate[attrname] or {
+    co = cocreate(_renderAttribute),
+    particles = _newParticleSource()
+  }
+  local particles = attrstate.particles
+  local data = {actor, attrname, particles}
+  local _, rawlvl, offset, percent, rise = assert(coresume(attrstate.co, data))
+  actorstate[attrname] = attrstate
+  _states[actor] = actorstate
+
+
+  -- check if attribute is modified
+  local lvl = actor:getAttribute(attrname) - offset
+  local diff = lvl - rawlvl
+  local color = (diff > 0 and {0, 0.7, 1}) or
+                (diff < 0 and {1, 0.8, 0.2}) or
+                COLORS.NEUTRAL
+
+  -- render bar
+  particles:update(_dt)
+  _font:set()
+  g.setColor(COLORS.NEUTRAL)
+  g.printf(
+    {
+      COLORS.NEUTRAL,
+      ("%s: "):format(attrname),
+      rise and COLORS.NEUTRAL or color,
+      ("%02d"):format(lvl)
+    },
+    0, 0, _barwidth, "left"
+  )
+  g.translate(0, 32)
+  g.push()
+  if not rise then
+    g.setColor(COLORS.EMPTY)
+  end
+  g.rectangle("fill", 0, 0, _barwidth, 16)
+  if not rise then 
+    g.setColor(COLORS[attrname])
+  end
+  g.rectangle("fill", 0, 0, percent*_barwidth, 16)
+  g.setColor(COLORS.NEUTRAL)
+  g.pop()
+  --_rot = fmod(_rot - 0.5*pi*_dt, 2*pi)
+  g.draw(particles, 48, -_font:getHeight()/2, _rot)
```

Comment:
The `_rot` variable doesn't seem to be used anywhere, and is always equal to zero.